File: app/src/main/java/ru/radiationx/anilibria/ui/fragments/releases/ReleasesFragment.java
package ru.radiationx.anilibria.ui.fragments.releases;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.arellomobile.mvp.presenter.InjectPresenter;

import java.util.ArrayList;

import ru.radiationx.anilibria.R;
import ru.radiationx.anilibria.data.api.releases.ReleaseItem;
import ru.radiationx.anilibria.ui.fragments.BaseFragment;

/* Created by radiationx on 05.11.17. */

public class ReleasesFragment extends BaseFragment implements ReleasesView, ReleasesAdapter.ItemListener {
    private RecyclerView recyclerView;
    private SwipeRefreshLayout refreshLayout;
    private ReleasesAdapter adapter;
    @InjectPresenter/*(tag = "ReleasesTag", type = PresenterType.GLOBAL)*/
            ReleasesPresenter presenter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.e("SUKA", "onCreate: " + this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.e("SUKA", "onDestroy: " + this);
    }

    @Override
    public void onCreateView(LayoutInflater inflater, @Nullable Bundle savedInstanceState) {
        baseInflateFragment(inflater, R.layout.fragment_releases);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        Log.e("SUKA", "onViewCreated");
        toolbar.setTitle(R.string.fragment_title_releases);
        Log.e("SUKA", "onViewCreated title " + toolbar.getTitle());
        refreshLayout = view.findViewById(R.id.swipe_refresh);
        recyclerView = view.findViewById(R.id.recycler_view);
        adapter = new ReleasesAdapter(getMvpDelegate());
        adapter.setListener(this);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext()));

        refreshLayout.setOnRefreshListener(() -> {
            Log.e("SUKA", "setOnRefreshListener");
            presenter.refreshReleases();
        });
        toolbar.getMenu().add("Поиск")
                .setIcon(R.drawable.ic_toolbar_search)
                .setOnMenuItemClickListener(item -> {
                    Toast.makeText(getContext(), "Временно не поддерживается", Toast.LENGTH_SHORT).show();
                    return false;
                })
                .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
    }

    @Override
    public void showReleases(ArrayList<ReleaseItem> releases) {
        Log.e("SUKA", "showReleases");
        adapter.bindItems(releases);
    }

    @Override
    public void insertMore(ArrayList<ReleaseItem> releases) {
        Log.e("SUKA", "insertMore");
        adapter.insertMore(releases);
    }

    @Override
    public void onLoadMore() {
        Log.e("SUKA", "onLoadMore");
        presenter.loadMore();
    }

    @Override
    public void setRefreshing(boolean refreshing) {
        refreshLayout.setRefreshing(refreshing);
    }

    @Override
    public void onItemClick(ReleaseItem item) {
        presenter.onItemClick(item);
    }

    @Override
    public boolean onItemLongClick(ReleaseItem item) {
        return presenter.onItemLongClick(item);
    }
}

File: app/src/main/java/ru/radiationx/anilibria/ui/fragments/releases/ReleasesFragment.kt
package ru.radiationx.anilibria.ui.fragments.releases

import android.os.Bundle
import android.support.v4.app.SharedElementCallback
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import kotlinx.android.synthetic.main.fragment_main_base.*
import kotlinx.android.synthetic.main.fragment_releases.*
import ru.radiationx.anilibria.App
import ru.radiationx.anilibria.R
import ru.radiationx.anilibria.data.api.models.ReleaseItem
import ru.radiationx.anilibria.ui.common.RouterProvider
import ru.radiationx.anilibria.ui.fragments.BaseFragment
import ru.radiationx.anilibria.ui.fragments.release.ReleaseFragment
import java.util.*

/* Created by radiationx on 05.11.17. */

class ReleasesFragment : BaseFragment(), ReleasesView, ReleasesAdapter.ItemListener {
    override val layoutRes: Int = R.layout.fragment_releases
    private var adapter: ReleasesAdapter = ReleasesAdapter()

    companion object {
        const val SUKA = "su4ara_rv"
    }

    @InjectPresenter
    lateinit var presenter: ReleasesPresenter

    @ProvidePresenter
    fun provideReleasesPresenter(): ReleasesPresenter {
        return ReleasesPresenter(App.injections.releasesRepository,
                (parentFragment as RouterProvider).router)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        refreshLayout.setOnRefreshListener { presenter.refreshReleases() }

        recyclerView.apply {
            adapter = this@ReleasesFragment.adapter
            layoutManager = LinearLayoutManager(recyclerView.context)
        }

        adapter.setListener(this)
        toolbar.apply {
            title = getString(R.string.fragment_title_releases)
            menu.add("Поиск")
                    .setIcon(R.drawable.ic_toolbar_search)
                    .setOnMenuItemClickListener({
                        presenter.openSearch()
                        //Toast.makeText(context, "Временно не поддерживается", Toast.LENGTH_SHORT).show()
                        false
                    })
                    .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
        }
    }

    override fun onBackPressed(): Boolean {
        presenter.onBackPressed()
        return true
    }

    override fun setEndless(enable: Boolean) {
        adapter.endless = enable
    }

    override fun showReleases(releases: ArrayList<ReleaseItem>) {
        Log.e("SUKA", "showReleases")
        adapter.bindItems(releases)
    }

    override fun insertMore(releases: ArrayList<ReleaseItem>) {
        Log.e("SUKA", "insertMore")
        adapter.insertMore(releases)
    }

    override fun onLoadMore() {
        Log.e("SUKA", "onLoadMore")
        presenter.loadMore()
    }

    override fun setRefreshing(refreshing: Boolean) {
        refreshLayout.isRefreshing = refreshing
    }

    var lsastpost = -1
    override fun onItemClick(item: ReleaseItem, position: Int) {
        lsastpost = position
        presenter.onItemClick(item)
    }

    lateinit var lastView: View

    override fun onItemClick(position: Int, view: View) {
        Log.e("SUKA", "ONITEMCLIC " + view)
        this.lastView = view;
    }

    override fun onItemLongClick(item: ReleaseItem): Boolean {
        return presenter.onItemLongClick(item)
    }


    fun getSharedView(): View {
        val viewItem = recyclerView.getLayoutManager().findViewByPosition(lsastpost)
        val icon = viewItem.findViewById<ImageView>(R.id.item_image)
        //lastView.transitionName = ReleaseFragment.TRANSACTION
        return lastView
    }
}
